File: apps/pwa/src/server/api/get_musicbill.ts
import { MusicType } from '@/constants/music';
import { prefixServerOrigin } from '@/global_states/server';
import { request } from '..';

interface User {
  id: string;
  nickname: string;
  avatar: string;
}

interface Response {
  id: string;
  cover: string;
  name: string;
  public: boolean;
  createTimestamp: number;
  owner: User;
  sharedUserList: (User & {
    accepted: boolean;
  })[];
  musicList: {
    id: string;
    type: MusicType;
    name: string;
    aliases: string[];
    cover: string;
    coverThumbnail?: string;
    asset: string;
    singers: {
      id: string;
      name: string;
      aliases: string[];
      avatar: string;
    }[];
    lyricists: {
      id: string;
      name: string;
      aliases: string[];
      avatar: string;
    }[];
  }[];
}

/**
 * 获取乐单
 * @author mebtte<i@mebtte.com>
 */
async function getMusicbill(id: string) {
  const musicbill = await request<Response>({
    path: '/api/musicbill',
    params: { id },
    withToken: true,
  });
  return {
    ...musicbill,
    cover: prefixServerOrigin(musicbill.cover),
    owner: {
      ...musicbill.owner,
      avatar: prefixServerOrigin(musicbill.owner.avatar),
    },
    sharedUserList: musicbill.sharedUserList.map((u) => ({
      ...u,
      avatar: prefixServerOrigin(u.avatar),
    })),
    musicList: musicbill.musicList.map((m) => ({
      ...m,
      cover: prefixServerOrigin(m.cover),
      coverThumbnail: prefixServerOrigin(m.coverThumbnail ?? ''),
      asset: prefixServerOrigin(m.asset),
      lyricists: m.lyricists ?? [],
    })),
  };
}

export default getMusicbill;

File: apps/flutter/lib/server/api/get_musicbill.dart
import 'package:cicada/models/music.dart';
import '../../utils/prefix_server_origin.dart';
import '../request.dart';

class Musicbill {
  String name;
  String? cover;
  bool isPublic;
  bool isShared;
  List<Music> musicList;

  Musicbill({
    required this.name,
    required this.cover,
    required this.isPublic,
    required this.isShared,
    required this.musicList,
  });

  factory Musicbill.fromJson(Map<String, dynamic> json) => Musicbill(
    name: json['name'],
    cover: prefixServerOrigin(json['cover']),
    isPublic: json['public'] == true,
    isShared: (json['sharedUserList'] as List<dynamic>? ?? const []).isNotEmpty,
    musicList: (json['musicList'] as List<dynamic>)
        .map((json) => Music.fromJson(json))
        .toList(),
  );
}

Future<Musicbill> getMusicbill({required String id}) async {
  final responseData = await httpGet(
    path: "/api/musicbill",
    query: {"id": id},
    withToken: true,
  );
  return Musicbill.fromJson(responseData);
}
